### PR TITLE
completed technical learning task week 1 FE-58

### PR DESCRIPTION
**Pull request by 5oappy**
## Describe your changes
implemented a base refactored version of 'unavailability_update' function from 'volunteer_unavailability.py',  compartmentalising the function into a new class 'unavailability_update.py' with new endpoints to familiarise with code base and api framework. Tested briefly using postman.

image

image

## Issue ticket number and link
jira FE-58 "Handle the Create operation to add a new unavailability record for a volunteer in the database."

### CODE REVIEW

File: controllers/unavailability_update.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                    args['end'],
+                    args['periodicity']
+                )
+                response = {
+                    "eventId": eventId,
+                    "success": eventId is not None
+                }
+                return jsonify(response)
+        except Exception as e:
+            return jsonify({"error": str(e), "success": False})
```
[the remark this answers is not in the document]
Comment (BenMMcLean) [reply controllers/unavailability_update.py:49]: Don't reveal the stack trace here

File: controllers/unavailability_update.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+volunteer_unavailability_update_bp = Blueprint('unavailability_update', __name__)
+api = Api(volunteer_unavailability_update_bp,'/update')
+api.add_resource(CreateNewUnavailabilityEvent_v2, '/createUnavailableEventV2')
```
Comment (BenMMcLean) [inline controllers/unavailability_update.py:54]: These endpoint names don't follow standard [REST endpoint naming conventions](https://restfulapi.net/resource-naming/)

File: controllers/unavailability_update.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenMMcLean) [reply controllers/unavailability_update.py:49]: The security bot is right here, you shouldn't reveal the stack trace. Just return a 500.

File: controllers/unavailability_update.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                    "success": eventId is not None
+                }
+                return jsonify(response)
+        except Exception as e:
+            return jsonify({"error": str(e), "success": False})
+
+
+volunteer_unavailabilities_bp = Blueprint('unavailabilities', __name__)
+api = Api(volunteer_unavailabilities_bp, '/unavailabilities')
+api.add_resource(CreateNewUnavailabilityEvent_v2, '/create')
```
Comment (BenMMcLean) [inline controllers/unavailability_update.py:54]: This is close, but I think "/unavailability" would be more appropriate. "create" is implied by the POST operation. Even better would be "/v2/volunteer/<user_id>/unavailability", because it clearly defines the relationship between unavailability and the user.

File: controllers/unavailability_update.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                    session,
+                    args['userId'],
+                    args['title'],
+                    args['start'],
+                    args['end'],
+                    args['periodicity']
+                )
+                response = {
+                    "eventId": eventId,
+                    "success": eventId is not None
```
Comment (BenMMcLean) [inline controllers/unavailability_update.py:45]: Please don't return a success field. You should be returning a http status of 200 if you are successful, 400 if the request failed for a non-error reason, and 500 if it failed for an error reason. This feedback also applies to line 49

**Review by BenMMcLean — CHANGES_REQUESTED** [review]
I'm quite happy with this, but I just want some changes to how the response is delivered, as well as seeing the endpoint in the form "/v2/volunteer/<user_id>/unavailability"

**Comment by 5oappy** [comment]
implemented everything 
image

File: controllers/volunteer_unavailability.py [hunk 5]
```diff
@@ -13,8 +13,8 @@
 create_parser.add_argument('userId', type=int, required=True)
```
Comment (BenMMcLean) [inline controllers/volunteer_unavailability.py:13]: Please revert these changes, they aren't relevant

File: controllers/unavailability_update.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+from flask_restful import reqparse, Resource, fields, marshal_with, Api, inputs, abort
+from domain import session_scope
+from repository.unavailability_repository import *
+
+from services.jwk import requires_auth
+
+# Request parser for creating a new unavailability event
+create_parser = reqparse.RequestParser()
+create_parser.add_argument('userId', type=int, required=True)
```
Comment (BenMMcLean) [inline controllers/unavailability_update.py:11]: Remove userId from the body. You're getting it from the URL

File: controllers/unavailability_update.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
+from flask import Blueprint, request, jsonify
```
Comment (BenMMcLean) [inline controllers/unavailability_update.py:1]: This file should be in v2/unavailability, and it should be called api. Please merge in main, integrate your file into the existing api.py and replicate the changes I made in my model PR #161. Everything should be the same except for the line ``v2_api.add_resource(VolunteerUnavailabilityV2, '/v2/volunteers/', '/v2/volunteers/<user_id>/unavailability/<event_id>')``. You'll want yours to be registered with ``v2_api.add_resource(CreateVolunteerUnavailabilityV2, '/v2/volunteers/', '/v2/volunteers/<user_id>/unavailability')``.

File: controllers/unavailability_update.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+# Request parser for creating a new unavailability event
+create_parser = reqparse.RequestParser()
+create_parser.add_argument('userId', type=int, required=True)
+create_parser.add_argument('title', type=str, required=True)
+create_parser.add_argument('periodicity', type=int, required=True)
+create_parser.add_argument('start', type=inputs.datetime_from_iso8601, required=True)
+create_parser.add_argument('end', type=inputs.datetime_from_iso8601, required=True)
+
+
+class CreateNewUnavailabilityEvent_v2(Resource):
```
Comment (BenMMcLean) [inline controllers/unavailability_update.py:18]: This should be named CreateUnavailabilityEventV2 (mixing title and snake case isn't good practice).

File: controllers/unavailability_update.py (hunk 7 above)
Comment (BenMMcLean) [reply controllers/unavailability_update.py:1]: It's fine if you're not quite sure what your doing or why. Ask me for help if needed and I'll cover it in more detail whenever I meet with the backend team :)

**Review by BenMMcLean — CHANGES_REQUESTED** [review]

**Comment by 5oappy** [comment]
Okay so I've tossed everything over into the correct directory, re-wrote most of it including the end points to match the style of other functions. Its been tested briefly in postman and everything checks out from what I can see. 

 Had to delete unavailability_read from v2/__init__.py as it no longer exist but I think it was over looked when merged into main, it poped up originally when I fetched and rebased the new additions in main to my branch as it was behind due to either the new approved PRs or changes you've made to the main branch.
 
 I think that should be everything..... I hope

**Review by BenMMcLean — DISMISSED** [review]
Good work. I'll push a couple of changes and then merge

**Review by BenMMcLean — APPROVED** [review]
